### PR TITLE
Fix pivoted qr jvp rule, which was broken when vmapped

### PR DESCRIPTION
**Pull request by adconner**
The old qr jvp rule was broken when vmapped, see the following test:

```python
import jax
import jax.numpy as jnp
from functools import partial

m = jnp.arange(18,dtype=jnp.float64).reshape(2,3,3)
mt = jnp.arange(18,36,dtype=jnp.float64).reshape(2,3,3)

res1 = jax.jvp(jax.vmap(partial(jax.scipy.linalg.qr, pivoting=True, mode='economic')), (m,), (mt,))
res2 = jax.vmap(lambda m,mt: jax.jvp(partial(jax.scipy.linalg.qr, pivoting=True, mode='economic'), (m,), (mt,) )) ((m,), (mt,))
```

The fix is a single line.

### CODE REVIEW

File: jax/_src/lax/linalg.py [hunk 1]
```diff
@@ -1868,7 +1868,7 @@ def qr_jvp_rule(primals, tangents, *, pivoting, full_matrices, use_magma):
     raise NotImplementedError(
       "Unimplemented case of QR decomposition derivative")
   if pivoting:
-    dx = dx[..., p[0]]
+    dx = np.take_along_axis(dx, p[0][..., None, :])
```
[the remark this answers is not in the document]
Comment (adconner) [reply jax/_src/lax/linalg.py]: `np` here refers to `jax.numpy`, which handles tracers. The `axis` argument of `take_along_axis` defaults to `-1` since August 2025.

[the remark this answers is not in the document]
Comment (tttc3) [reply jax/_src/lax/linalg.py]: Thanks for fixing my mistake here! If it's helpful, here's a numpy-free alternative based on `_lu_pivots_body_fn`:

```python
*batch_dims, m, n = x.shape
if m < n or (full_matrices and m != n):
  raise NotImplementedError(
    "Unimplemented case of QR decomposition derivative")
if pivoting:
  permute_fn = lambda dx, p: dx[..., p]
  for _ in range(len(batch_dims)):
    permute_fn = api.vmap(permute_fn)
  dx = permute_fn(dx, p[0])
```
I recall there being a reason for avoiding numpy functions in these JVPs, though I'm not fully sure of the details. Someone else might be able to chime in with more context!

[the remark this answers is not in the document]
Comment (adconner) [reply jax/_src/lax/linalg.py]: Good idea to mirror how pivoted lu does it. Changed

File: jax/_src/lax/linalg.py [hunk 2]
```diff
@@ -1863,12 +1863,15 @@ def qr_jvp_rule(primals, tangents, *, pivoting, full_matrices, use_magma):
...
-  *_, m, n = x.shape
+  *batch_dims, m, n = x.shape
   if m < n or (full_matrices and m != n):
     raise NotImplementedError(
       "Unimplemented case of QR decomposition derivative")
   if pivoting:
-    dx = dx[..., p[0]]
+    permute_fn = lambda dx, p: dx[..., p]
+    for _ in range(len(batch_dims)):
+      permute_fn = api.vmap(permute_fn)
```
Comment (jakevdp) [inline jax/_src/lax/linalg.py:1873]: I think there's an issue here – here `batch_dims` may not contain the leading dimensions, so vmapping over the leading dimensions might produce an incorrect result

**Review by jakevdp — COMMENTED** [review]
Hey, sorry we lost track of this. I'm not totally clear on what's going on here, but I have one comment below.

Do you have an example of problematic inputs that we could turn into a regression test?

File: jax/_src/lax/linalg.py (hunk 2 above)
Comment (adconner) [reply jax/_src/lax/linalg.py:1873]: I'm not sure I understand the problem. `batch_dims` is set on line 1866 to contain the leading dimensions
```python
  *batch_dims, m, n = x.shape
```
I'm batching in the same way as in `_lu_jvp_rule` (line 1527). I could simplify to
```python
for _ in batch_dims:
      permute_fn = api.vmap(permute_fn)
```
if you prefer.

**Comment by adconner** [comment]
> Do you have an example of problematic inputs that we could turn into a regression test?

The example given in the PR raises an exception without the fix (both lines giving `res1` and `res2`)
```python
import jax
import jax.numpy as jnp
from functools import partial

m = jnp.arange(18,dtype=jnp.float64).reshape(2,3,3)
mt = jnp.arange(18,36,dtype=jnp.float64).reshape(2,3,3)

res1 = jax.jvp(jax.vmap(partial(jax.scipy.linalg.qr, pivoting=True, mode='economic')), (m,), (mt,))
res2 = jax.vmap(lambda m,mt: jax.jvp(partial(jax.scipy.linalg.qr, pivoting=True, mode='economic'), (m,), (mt,) )) ((m,), (mt,))
```
